src/rgw/rgw_rest_s3.cc

```diff
@@ -1477,7 +1477,13 @@ int RGWListBucket_ObjStore_S3::get_params()
     marker = s->info.args.get("marker");
   } else {
     marker.name = s->info.args.get("key-marker");
-    marker.instance = s->info.args.get("version-id-marker");
+    auto ver = s->info.args.get_optional("version-id-marker");
+    if (ver) {
+      marker.instance = *ver;
+    } else {
+      // compares larger than anything from gen_rand_obj_instance_name()
+      marker.instance = std::string{32, '~'};
```

Comment:
I do not like that this specialized knowledge is here rather than with the other code that handles instance names. The 32 is hard-coded. I envision a (sub-)class with static (and inline?) methods to generate randomized instance names, return the special "past end" name, and if you look below, test whether an instance names is the special "past end" name. Or better yet when given an instance name, provides its canonical string version, which would be the empty string in the case when the special "past end" name  is passed in.

RGW code spreads knowledge too far and wide. It's bad software engineering. I realize it's prevalent and I'm not blaming you.

Would you be willing to make these changes? If not, after your PR merges I'll submit a follow-up PR that does this.

Thanks for considering.